rfd/0078-login-rules.md

```diff
@@ -0,0 +1,703 @@
+`dict.remove` or `set.contains` as described above.
+We will need to add support for these in our
+[gravitational/predicate](https://github.com/gravitational/predicate)
+fork.
+
+^ Update: method support has been implemented
+[here](https://github.com/gravitational/predicate/pull/4), remaining changes
+can all happen in teleport.
+
+### Using transformed traits in roles
+
+Login rules transparently set the user's traits during login.
+To the rest of the cluster, including role templates, they will appear identical
+to normal traits which are typically reference by `{{external.<trait_name>}}`.
+
+### When login rules will be parsed and evaluated
+
+Login rules will be parsed and evaluated during each user login.
+This will occur after the SSO provider has returned its assertions/claims, and
+before Teleport maps these to internal Teleport roles via `attributes_to_roles` or
+`claims_to_roles` so that transformed traits can be used for this mapping.
+
+An eventual extension to login rules could also return the desired set of
+Teleport roles which the user should have.
+
+During login, the auth server will load all `login_rule` resources in the
+cluster, sort them by `priority` and `name`, and apply all of them in order.
+
+### Trusted clusters
+
+Since login rules will be evaluated during login and the resulting traits will
+be embedded in the user's certificates, they only need to be created in the root
+cluster and the transformed traits will be visible and usable in all leaf
+clusters.
+
+### Creating and modifying login rules
+
+Login rules will be a new backend resource `login_rule`.
+
+They should be written in a YAML file and created by the usual means of
+`tctl create resource.yaml`. `tctl get login_rule/example` can be used to
+fetch the current resource, and `tctl rm login_rule/example` can be used
+to delete the resource.
+
+### Validating and testing login rules
+
+A new `tctl` command will be introduced, which can be used to experiment with
+login rules and test what their effect will be before deploying them to
+your cluster.
+
+```bash
+$ tctl test login_rule \
+  --load login_rule1.yaml \
+  --load login_rule2.yaml \
+  --load-from-cluster \
+  --input_traits '{"groups": ["splunk"], "email": "nic@goteleport.com", "username": "nklaassen"}'
+```
+
+You can load multiple yaml files with `--load resource.yaml`, optionally load
+existing trait transforms from the cluster with `--load-from-cluster`, and
+provide a set of input traits to test with.
+
+The command will report any syntax errors, and will print the output traits for
+the given input.
+
+### Protobuf Definitions
+
+The `login_rule` resource and associated CRUD RPCs will be added to a new
+package `teleport/loginrule/v1`.
+
+```
+// loginrule.proto
+...
+
+// LoginRule is a resource to configure rules and logic which should run during
+// Teleport user login.
+message LoginRule {
+  // kind is a resource kind
```

Comment:
What is the kind for LoginRules? Does it support a multitude of kinds or only one?

I understand the fields here are added to make LoginRules conform to the "resources" interface, but I would consider alternatives to be conformant without adding superficial fields to the gRPC/proto APIs.

Reply:
the kind will be `login_rule` as it's shown in all the YAML examples. That's the only kind relevant for this type, it's used for YAML parsing, I'll see about keeping that out of the protobufs